src/quarto/ai/minimax.go

```diff
@@ -8,66 +8,143 @@ import (
+				if len(piecesList) > 1 && len(boxList) > 1 {
+					for j := 0; j < len(piecesList) ; j++ {
+						if node.State.Piece != piecesList[j] {
+							if node.State.Piece == 0 {
 								nextState := state.CopyState(node.State)
-								nextState.Grid[boxList[i].Y][boxList[i].X] = node.State.Piece
 								nextState.Piece = piecesList[j]
-								childNode := InitNode(nextState)
-								node.Childs = append(node.Childs, AppendChildNodes(childNode, depth-1, quit ))
+								childNode := InitNode(nextState, !node.MyNode)
+								childPiecesList := grid.GetListPieceMinusPiece(piecesList, nextState.Piece)
+								node.Childs = append(node.Childs, AppendChildNodes(childNode, depth-1, childPiecesList, boxList, quit))
+							} else {
+								for i := 0; i < len(boxList); i++ {
+									nextState := state.CopyState(node.State)
+									nextState.Grid[boxList[i].Y][boxList[i].X] = node.State.Piece
+									nextState.Move = [2]int{boxList[i].Y, boxList[i].X}
+									nextState.Piece = piecesList[j]
+									childNode := InitNode(nextState, !node.MyNode)
+									if grid.IsWinningPosition(boxList[i].X, boxList[i].Y, node.State.Grid, node.State.Piece) {
+										if (childNode.MyNode) {
+											childNode.Value = WinningLeafValue
+										} else {
+											childNode.Value = LoosingLeafValue
+										}
+										node.Childs = append(node.Childs, childNode)
+									} else {
+										childPiecesList := grid.GetListPieceMinusPiece(piecesList, nextState.Piece)
+										childBoxList := grid.GetListBoxMinusPoint(boxList, boxList[i])
+										node.Childs = append(node.Childs, AppendChildNodes(childNode, depth-1, childPiecesList, childBoxList, quit ))
+									}
+								}
 							}
 						}
 					}
+				} else if len(piecesList) == 1 && len(boxList) == 1 {
+					nextState := state.CopyState(node.State)
+					nextState.Grid[boxList[0].Y][boxList[0].X] = node.State.Piece
+					nextState.Move = [2]int{boxList[0].Y, boxList[0].X}
+					nextState.Piece = 0
+					childNode := InitNode(nextState, !node.MyNode)
+
+					if grid.IsWinningPosition(boxList[0].X, boxList[0].Y, node.State.Grid, node.State.Piece) {
+						if (childNode.MyNode) {
+							childNode.Value = WinningLeafValue
+						} else {
+							childNode.Value = LoosingLeafValue
+						}
+					} else {
+						childNode.Value = WinningLeafValue
+					}
+					node.Childs = append(node.Childs, childNode)
 				}
 			}
+			if (node.Value == UndefinedValue) {
+				node.Value = CalculateNodeValue(node)
+			}
 			return node
 	}
 }
 
-func PrintTree(node StateNode, depth int) {
+func ChooseBestChildState(node StateNode) state.State {
+	value := LoosingLeafValue + 1
+	bestState := state.State{}
+	for i := 0; i < len(node.Childs); i++ {
+		if value > node.Childs[i].Value && node.Childs[i].Value != UndefinedValue {
+			value = node.Childs[i].Value
+			bestState = node.Childs[i].State
+		}
+	}
+	return bestState
+}
+
+func CalculateNodeValue(node StateNode) int {
```

Comment:
I don't like this function but I don't find a prettier version ^^

Reply:
Actually, the same ...